module_utils/certificate/providers/base.py

```diff
@@ -430,6 +430,13 @@ def __eq__(self, other):
         self_info = {k: v for k, v in self.cert_data.items() if v}
         other_info = {k: v for k, v in other.cert_data.items() if v}
 
+        self._module.debug(
+            "Preparing CertificateProxy objects for comparison: "
+            "key_usage attribute will be sorted."
+        )
+        self_info.get("key_usage", []).sort()
+        other_info.get("key_usage", []).sort()
```

Comment:
This modifies the data in `self.cert_data['key_usage']` and in `other.cert_data['key_usage']` - that is, it modifies the data in-place as a side effect of the comparison:
```python
>>> self_cert_data = {'keyA':'valueA', 'key_usage':['C', 'B', 'A']}
>>> other_cert_data = {'keyA':'valueA', 'key_usage':['B', 'A', 'C']}
>>> self_info = {k: v for k, v in self_cert_data.items() if v}
>>> other_info = {k: v for k, v in other_cert_data.items() if v}
>>> self_info['key_usage'] is self_cert_data['key_usage']
True
>>> self_info.get("key_usage", []).sort()
>>> other_info.get("key_usage", []).sort()
>>> self_cert_data
{'keyA': 'valueA', 'key_usage': ['A', 'B', 'C']}
>>> other_cert_data
{'keyA': 'valueA', 'key_usage': ['A', 'B', 'C']}
```

Reply:
it seems that a list is not the best data type to represent this information, could one use a  [set](https://docs.python.org/2/library/stdtypes.html#set-types-set-frozenset)?

Reply:
That's true @richm. That side effect was unintended. We can use a copy of the key instead of just the using the key to avoid that.

@pcahyna using a sounds like a good idea and would probably work.

Reply:
@pcahyna that's a great idea to use sets.

I modified the patch using sets for the key_usage parameter.